src/vs/workbench/contrib/tasks/test/browser/taskTerminalStatus.test.ts

```diff
@@ -0,0 +1,85 @@
+	private readonly _onDidStateChange: Emitter<TaskEvent> = new Emitter();
+	public get onDidStateChange(): Event<TaskEvent> {
+		return this._onDidStateChange.event;
+	}
+	public triggerStateChange(event: TaskEvent): void {
+		this._onDidStateChange.fire(event);
+	}
+}
+
+class TestTerminal implements Partial<ITerminalInstance> {
+	statusList: TerminalStatusList = new TerminalStatusList(new TestConfigurationService());
+}
+
+class TestTask extends CommonTask {
+	protected getFolderId(): string | undefined {
+		throw new Error('Method not implemented.');
+	}
+	protected fromObject(object: any): Task {
+		throw new Error('Method not implemented.');
+	}
+}
+
+class TestProblemCollector implements Partial<AbstractProblemCollector> {
+
+}
+
+const instantiationService = new TestInstantiationService();
+suite('Task Terminal Status', () => {
+	let taskService: TestTaskService;
+	let taskTerminalStatus: TaskTerminalStatus;
+	let testTerminal: ITerminalInstance;
+	let testTask: Task;
+	let problemCollector: AbstractProblemCollector;
+	setup(() => {
+		taskService = new TestTaskService();
+		taskTerminalStatus = instantiationService.createInstance(TaskTerminalStatus, taskService);
+		testTerminal = instantiationService.createInstance(TestTerminal);
+		testTask = instantiationService.createInstance(TestTask);
+		problemCollector = instantiationService.createInstance(TestProblemCollector);
+	});
+	test('Should add failed status when there is an exit code on task end', async () => {
+		taskTerminalStatus.addTerminal(testTask, testTerminal, problemCollector);
+		taskService.triggerStateChange({ kind: TaskEventKind.ProcessStarted });
+		ok(assertStatus(testTerminal.statusList, ACTIVE_TASK_STATUS));
+		taskService.triggerStateChange({ kind: TaskEventKind.Inactive });
+		ok(assertStatus(testTerminal.statusList, SUCCEEDED_TASK_STATUS));
+		taskService.triggerStateChange({ kind: TaskEventKind.End, exitCode: 1 });
+		ok(assertStatus(testTerminal.statusList, FAILED_TASK_STATUS));
+	});
+	test('Should add active status when a non-background task is run for a second time in the same terminal', async () => {
+		taskTerminalStatus.addTerminal(testTask, testTerminal, problemCollector);
+		taskService.triggerStateChange({ kind: TaskEventKind.ProcessStarted });
+		ok(assertStatus(testTerminal.statusList, ACTIVE_TASK_STATUS));
+		taskService.triggerStateChange({ kind: TaskEventKind.Inactive });
+		ok(assertStatus(testTerminal.statusList, SUCCEEDED_TASK_STATUS));
+		taskService.triggerStateChange({ kind: TaskEventKind.ProcessStarted, runType: TaskRunType.SingleRun });
+		ok(assertStatus(testTerminal.statusList, ACTIVE_TASK_STATUS));
+		taskService.triggerStateChange({ kind: TaskEventKind.Inactive });
+		ok(assertStatus(testTerminal.statusList, SUCCEEDED_TASK_STATUS));
+	});
+});
+
+function assertStatus(actual: ITerminalStatusList, expected: ITerminalStatus): boolean {
+	if (actual.statuses.length !== 1) {
+		return false;
+	}
+	return actual.statuses[0].id === expected.id;
```

Comment:
I also wanted to check the `severity` as well, but for some reason, that is not correctly updated for the first test unless I add a `setTimeout`